src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -101,60 +237,862 @@ def cert_issuer(cert):
+                crypto.FILETYPE_PEM,
+                certificate
+            )
+        except crypto.Error:
+            return {}
+        else:
+            cert_info = {
+                'country': cert.get_subject().C,
+                'state': cert.get_subject().ST,
+                'city': cert.get_subject().L,
+                'organization': cert.get_subject().O,
+                'common': cert.get_subject().CN,
+                'san': cert.get_subject().subjectAltName,
+                'email': cert.get_subject().emailAddress,
+            }
+
+            signature_algorithm = cert.get_signature_algorithm().decode()
+            m = re.match('^(.+)[Ww]ith', signature_algorithm)
+            if m:
+                cert_info['digest_algorithm'] = m.group(1).upper()
+
+            return cert_info
+
+    @accepts(
+        Str('buffer', required=True),
+        Str('passphrase', default=None)
+    )
+    async def export_private_key(self, buffer, passphrase):
+        key = await self.load_private_key(buffer, passphrase)
+        if key:
+            return crypto.dump_privatekey(
+                crypto.FILETYPE_PEM,
+                key,
+                passphrase=passphrase.encode() if passphrase else None
+            ).decode()
+
+    @accepts(
+        Str('buffer', required=True),
+        Str('passphrase', default=None)
+    )
+    async def load_private_key(self, buffer, passphrase):
+        try:
+            return crypto.load_privatekey(
+                crypto.FILETYPE_PEM,
+                buffer,
+                passphrase=passphrase.encode() if passphrase else None
+            )
+        except crypto.Error:
+            return None
+
+    @accepts(
+        Int('certificate_id', required=True)
+    )
+    async def get_fingerprint(self, certificate_id):
+        certificate_list = await self.query(filters=[('id', '=', certificate_id)])
+        if len(certificate_list) == 0:
+            return None
+        else:
+            cert_certificate = certificate_list[0]['certificate']
+
+        # getting fingerprint of certificate
+        try:
+            certificate = crypto.load_certificate(
+                crypto.FILETYPE_PEM,
+                cert_certificate
+            )
+        except Exception:
+            return None
+        else:
+            return certificate.digest('sha1').decode()
+
+    @accepts(
+        Dict(
+            'certificate_cert_info',
+            Int('key_length'),
+            Int('lifetime', required=True),
+            Str('country', required=True),
+            Str('state', required=True),
+            Str('city', required=True),
+            Str('organization', required=True),
+            Str('common', required=True),
+            Str('email', validators=[Email()], required=True),
+            Str('serial', required=False),
+            Str('digest_algorithm', enum=['SHA1', 'SHA224', 'SHA256', 'SHA384', 'SHA512']),
+            List('san', items=[Str('san')], default=[]),
```

Comment:
empty list ([]) cannot be a default, single its not a singleton and object is carried away between calls.